pubsub_nats.go

```diff
@@ -136,15 +155,37 @@ func (p *natsPubSub) Subscribe(pubs chan *Publication, errors chan error, topic
 		}
 
 		if m.Reply != "" {
-
-			resp := ackMessage
-			if config.replier != nil {
-				resp = config.replier(m)
-			}
-
-			if err := p.client.Publish(m.Reply, resp); err != nil {
-				zap.L().Error("Unable to send requested reply", zap.Error(err))
-				return
+			switch publication.ResponseMode {
+			// `ReplyWithACK` mode responds to the client right away, BEFORE the subscriber has had the opportunity
+			// to process the publication. Most consumers do message processing asynchronously and simply need
+			// to respond to the publisher with an ACK.
+			case ReplyWithACK:
+				if err := p.client.Publish(m.Reply, ackMessage); err != nil {
+					zap.L().Error("Failed to publish ACK reply. Message dropped.",
+						zap.Error(err),
+						zap.String("NATS subject", m.Reply))
+					return
+				}
+			// `ReplyWithPublication` mode is used in cases when the subscriber needs to do processing on the
+			// received publication PRIOR to publishing a response back. In such case, the received publication
+			// will have a write-only response channel set up so that the subscriber can send a response publication
+			// to whenever it is ready. The subscriber SHOULD attempt to respond ASAP as there is a client waiting
+			// for a response.
+			case ReplyWithPublication:
+				replyCh := make(chan *Publication)
+				publication.replyCh = replyCh
+				go responseHandler(m.Reply, replyCh)
+			// Finally, if client has configured a custom replier using the NATSOptSubscribeReplyer, use that to generate
+			// a response.
+			default:
+				if config.replier != nil {
```

Comment:
💡 We should get rid of this now. I just left it in for backwards compatibility. With the new request-reply workflow you can use the `ReplyWithPublication` if you expect a "custom" response message.

Reply:
Yes

Reply:
Done, it's gone now 🔥